runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java

```diff
@@ -18,62 +18,135 @@
+    executor.allowCoreThreadTimeOut(true);
+    this.maximumElementsOutstanding = maximumElementsOutstanding;
+    this.maximumBytesOutstanding = maximumBytesOutstanding;
+  }
+
+  // Before adding a Work to the queue, check that there are enough bytes of space or no other
+  // outstanding elements of work.
+  public void execute(Runnable work, long workBytes) {
+    monitor.enterWhenUninterruptibly(
+        new Guard(monitor) {
+          @Override
+          public boolean isSatisfied() {
+            return elementsOutstanding == 0
+                || (bytesAvailable() >= workBytes
+                    && elementsOutstanding < maximumElementsOutstanding);
+          }
+        });
+    executeLockHeld(work, workBytes);
   }
 
-  // Before adding a Runnable to the queue, acquire the semaphore.
-  @Override
-  public void execute(Runnable r) {
-    semaphore.acquireUninterruptibly();
-    super.execute(r);
+  public void forceExecute(Runnable work, long workBytes) {
+    monitor.enter();
+    executeLockHeld(work, workBytes);
   }
 
   // Forcibly add something to the queue, ignoring the length limit.
   public void forceExecute(Runnable r) {
-    semaphore.reducePermits(1);
-    super.execute(r);
+    forceExecute(r, 0);
+  }
+
+  public void shutdown() throws InterruptedException {
+    executor.shutdown();
+    if (!executor.awaitTermination(5, TimeUnit.MINUTES)) {
+      throw new RuntimeException("Work executor did not terminate within 5 minutes");
+    }
+  }
+
+  public boolean executorQueueIsEmpty() {
+    return executor.getQueue().isEmpty();
+  }
+
+  public String summaryHtml() {
+    monitor.enter();
+    try {
+      StringBuilder builder = new StringBuilder();
+      builder.append("Worker Threads: ");
+      builder.append(executor.getPoolSize());
+      builder.append("/");
+      builder.append(executor.getMaximumPoolSize());
+      builder.append("<br>/n");
+
+      builder.append("Active Threads: ");
+      builder.append(executor.getActiveCount());
+      builder.append("<br>/n");
+
+      builder.append("Work Queue Size: ");
+      builder.append(elementsOutstanding);
+      builder.append("/");
+      builder.append(maximumElementsOutstanding);
+      builder.append("<br>/n");
+
+      builder.append("Work Queue Bytes: ");
+      builder.append(bytesOutstanding);
+      builder.append("/");
+      builder.append(maximumBytesOutstanding);
+      builder.append("<br>/n");
+
+      return builder.toString();
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  private void executeLockHeld(Runnable work, long workBytes) {
+    Runnable workRunnable =
+        () -> {
+          try {
+            work.run();
+          } finally {
+            monitor.enter();
+            --elementsOutstanding;
+            bytesOutstanding -= workBytes;
+            monitor.leave();
+          }
+        };
+    try {
+      bytesOutstanding += workBytes;
+      ++elementsOutstanding;
+      executor.execute(workRunnable);
```

Comment:
seems like you should reduce monitor scope to just increment counters. Then make runnable and call execute outside

Might deadlock otherwise if forceexecutes have pushed the executor queue over maxelements and execute blocks queueing with monitor held.  Then work won't be able to enter and decrement when done and threads in executor will be tied up.